sshd-core/src/main/java/org/apache/sshd/common/session/helpers/AbstractSession.java

```diff
@@ -1334,9 +1362,19 @@ protected void decode() throws Exception {
                 assert decoderBuffer.rpos() == Integer.BYTES;
                 int macSize = (inMac != null) ? inMacSize : 0;
                 // Check if the packet has been fully received
-                if (decoderBuffer.available() >= (decoderLength + macSize)) {
+                if (decoderBuffer.available() >= (decoderLength + macSize + authLen)) {
                     byte[] data = decoderBuffer.array();
-                    if (etmMode) {
+                    if (authMode) {
+                        final BaseGCMCipher cipher = (BaseGCMCipher) inCipher;
+                        cipher.incrementIV();
+                        // RFC 5647: packet length encoded in additional data and unencrypted
+                        int off = decoderBuffer.rpos();
+                        cipher.updateAAD(data, off - Integer.BYTES, Integer.BYTES);
+                        cipher.update(data, off, decoderLength);
+                        cipher.doFinal(data, Integer.BYTES + decoderLength);
```

Comment:
Finally, I figured out one of the problems. This use of the Cipher class is incorrect. When verifying an AT, that should be processed in `update` rather than `doFinal`.